presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -747,9 +753,24 @@ public synchronized Map<String, Optional<Partition>> getPartitionsByNames(Metast
         }
         Map<String, Optional<Partition>> delegateResult = delegate.getPartitionsByNames(metastoreContext, databaseName, tableName, partitionNamesToQuery.build());
         resultBuilder.putAll(delegateResult);
+
+        cacheLastDataCommitTimes(delegateResult, databaseName, tableName);
+
         return resultBuilder.build();
     }
 
+    private void cacheLastDataCommitTimes(Map<String, Optional<Partition>> existingPartitions, String databaseName, String tableName)
+    {
+        List<DateTime> lastDataCommitTimes = existingPartitions.values().stream()
+                .filter(Optional::isPresent)
+                .map(partition -> partition.get().getLastDataCommitTime())
+                .map(commitTime -> new DateTime(TimeUnit.SECONDS.toMillis(commitTime)))
+                .limit(MAX_LAST_DATA_COMMIT_TIME_ENTRY)
+                .collect(toImmutableList());
+
+        this.lastDataCommitTimesForRead.put(new SchemaTableName(databaseName, tableName), lastDataCommitTimes);
```

Comment:
We should cap the total entries here

Reply:
How about <= 10K? (Assume the concurrency is <= 100)

Reply:
< 10K? (Concurrency <= 100)

Reply:
where is this capped?  also, i don't think this is generally used by multiple concurrent queries.  i believe it's used by multiple queries in a single transaction, so <100 should be fine.

Reply:
So a SemiTransactionalHiveMetastore instance is created for each individual query? I was waiting for the response; let me add the total cap now.